test/tests/selftest/pbs_param_dict.py

```diff
@@ -0,0 +1,125 @@
+
+
+from tests.selftest import *
+from ptl.utils.plugins.ptl_test_runner import PTLTestRunner
+
+
+class TestParamDict(TestSelf):
+    """
+    Test parsing the pbs_benchpress -p argument
+    """
+
+    def test_gpd(self):
+        """
+        Pass various -p param strings to __get_param_dictionary and check
+        that result is correct.
+        """
+        def try_one(runner, param):
+            """
+            Run get_param_dictionary to convert param string to dict
+            """
+            runner.param = param
+            try:
+                # Note that we call private routine directly to avoid
+                # possible side effects.
+                result = runner._PTLTestRunner__get_param_dictionary()
+            except ValueError:
+                result = None
+            return result
+
+        def compare(old, new, expect_none=False):
+            """
+            Compare new param dict to old
+            """
+            if new is None and not expect_none:
+                return 'Unexpected parse error'
+            if expect_none and new is not None:
+                return 'Should have failed'
+            all_keys = set(old.keys()) | set(new.keys())
+            diffs = []
+            for k in sorted(all_keys):
+                o = old.get(k, None)
+                n = new.get(k, None)
+                if o != n:
+                    diffs.append((k, n))
+            return diffs
+
+        def check_diffs(self, new, expected):
+            """
+            Validate changes between baseline and new param dict
+            """
+            if expected is None:
+                self.assertEqual(new, expected)
+                return
+            diffs = compare(self.base, new)
+            self.assertEqual(diffs, expected)
+
+        our_runner = PTLTestRunner()
+
+        # Create baseline dictionary
+        self.base = try_one(our_runner, '')
+
+        # Try simple key=value (boolean)
+        new = try_one(our_runner, 'mom_on_server=True')
+        check_diffs(self, new, [('mom_on_server', True)])
+
+        # Try one that generates a set of hostnames
+        new = try_one(our_runner, 'moms=foo')
+        check_diffs(self, new, [('moms', set(['foo']))])
+
+        # Test that unknown parameter is ignored
+        new = try_one(our_runner, 'nonsense')
+        check_diffs(self, new, [])
+
+        # Test more complicated host list
+        new = try_one(our_runner, 'comms=foo:bar@/path/to/bleem:baz')
+        check_diffs(self, new, [('comms', set(['foo', 'baz', 'bar']))])
+
+        # Test bad boolean arg
+        new = try_one(our_runner, 'mom_on_server=oops')
+        check_diffs(self, new, None)
+
+        # Test multiple options, one of which is default
+        new = try_one(our_runner,
+                      'server=foo2:bar,mom_on_server=y,comm_on_server=f')
+        expected = [('mom_on_server', True),
+                    ('servers', set(['foo2', 'bar']))]
```

Comment:
Why didn't comm_on_server come back in the list?

Reply:
That was part of the test. `comm_on_server=f `evaluates to False, which is its default value, so it doesn't show up as a difference.